### DISCUSSION TITLE
use Cache-control properties in Next.js getServersideProps

### DISCUSSION THREAD

**Discussion by wooseok123** [vercel/next.js#44902] [Help] [answered]
```typescript
import { GetServerSideProps, InferGetServerSidePropsType } from 'next'

export default function Index({
  time,
}: InferGetServerSidePropsType<typeof getServerSideProps>) {
  return (
    <main>
      <h1>SSR Caching with Next.js</h1>
      <time dateTime={time}>{time}</time>
    </main>
  )
}

export const getServerSideProps: GetServerSideProps<{ time: string }> = async ({
  res,
}) => {
  res.setHeader(
    'Cache-Control',
    'public, s-maxage=10, stale-while-revalidate=59'
  )

  return {
    props: {
      time: new Date().toISOString(),
    },
  }
}
```

https://github.com/vercel/next.js/tree/canary/examples/ssr-caching

I saw the example when caching in getServersideProps and I have some questions.

1) I heard that when we use getServersideProps, Next.js don't automatically cache cdn in contrast to the getStaticProps. But in the cache-control value, s-maxage is set 10, which is used for proxy or cdn. So, i think it is impossible to reuse the cached data. 

2) Actually, i can't understand the example above because when i clone this example and npm build / npm start, date data is always renewed every time i
 refresh the page. Then, what is the exact difference when there's no cache control value?  

-- Edited---
I think it only works on Vercel environment. Then, It means Vercel cache the data automatically by cdn even if it is defined in getServersideProps?

And is there a way to get cached data without using Vercel? 

Thank you!

**Comment by icyJoseph** [comment, marked as answer]
Right, so this where things leave the realm of frameworks and, enter the real of how does the internet actually, work, what things browsers say they do, what do they actually do, and what they never mention they do...

The main idea with setting a cache header is to tell CDN's that this content can be cacheable. Already we are implying the presence of three key players. A server, a CDN, and an entity that requests content. 

Usually the entity requesting the content is a browser. The problem is how browsers actually honour the cache-control.

If you load your page locally, and then back to the start page, and then forward to your page, you'll see it shows stale data. However, upon refreshing, browsers assume that user's actually perceived stale data, and want new data, so the browser does what it has to do to get fresh data. For example it may include:

```
Request Headers
Cache-Control: max-age=0
```

> Sometimes it might even send `no-cache`.

This post from the Chromium team sheds some light onto this,  https://blog.chromium.org/2017/01/reload-reloaded-faster-and-leaner-page_26.html:

> Users typically reload either because a page is broken or the content seems stale. The existing reload behavior usually solves broken pages, but stale content is inefficiently addressed by a regular reload, especially on mobile [...] To improve the stale content use case, Chrome now has a simplified reload behavior to only validate the main resource and continue with a regular page load. [...]

In general, I think all the major browsers take that stance. If the user reloaded the page, the user must think the data is stale, so the main entry point is refreshed. JSON, JavaScript, CSS, Fonts, etc, may be served from cache still though.

> And is there a way to get cached data without using Vercel?

By having your own CDN, for GSSP that is.  I think anyway.

**Reply by wooseok123** [reply]
Thank you for the detailed answer :)
